# M1L6 Data Challenge:  Missing Data

## Scenario

You are continuing your analysis of the "New York City Leading Causes of Death" dataset.  You've noticed that the Deaths and Death Rate columns contain some missing values, represented by periods ('.').  Missing data is a common issue in real-world datasets, and it's crucial to handle it appropriately to avoid biased or inaccurate conclusions.


For more information about the data (which is highly recommended) here is the [Link to the Data](https://data.cityofnewyork.us/Health/New-York-City-Leading-Causes-of-Death/jb7j-dtam/about_data)

## Objectives 

- Create missing value indicator columns.
- Identify the extent of missing data in the Deaths and Death Rate columns.
- Impute these missing values using appropriate strategies.



**Let's get started!**

### Step 1:  Import Pandas & Numpy

In [2]:
# Import Pandas 
import pandas as pd
import numpy as np

### Step 2: Load the dataset (csv file stored in the data folder) into a Pandas DataFrame. 

The file is called:  `nyc_causeofdeath.csv`


In [3]:
df = pd.read_csv('/Users/Marcy_Student/Desktop/marcy/DA2025_Lectures_Kevin/data/New_York_City_Leading_Causes_of_Death_20250605.csv')

### Step 3: Count up the number of missing values in 2 columns

a.  Use value_counts() to determine the number of missing values (represented by '.') in the `Deaths` and `Death Rate` columns.  Print the value counts for each column.  Be sure to set dropna=False as an argument within value_counts()

b.  Add a comment for the number of missig values (aka periods '.') are in each column


In [4]:
#Comment 1:  
print(df['Deaths'].value_counts())
#Comment 2:  
print(df['Death Rate'].value_counts())

Deaths
.       138
5        28
8        22
6        21
7        15
       ... 
261       1
171       1
1097      1
299       1
93        1
Name: count, Length: 465, dtype: int64
Death Rate
.        386
13         7
17.3       7
11.4       6
16.3       6
        ... 
19         1
6.9        1
91.6       1
35.1       1
170.3      1
Name: count, Length: 442, dtype: int64


Death has 138 missing values
Death Rate has 386 missing values. 

### Step 4:  Replace periods with NaN (not a number) and convert to numeric 

- a. Replace the '.' values in the `Deaths` and `Death Rate` columns with `np.nan`.
- b. Convert the `Deaths` and `Death Rate` columns to numeric.

This may take several lines of code

In [ ]:
df['Deaths'] = df['Deaths'].replace('.', np.nan)
df['Death Rate'] = df['Death Rate'].replace('.', np.nan)

df['Deaths'] = pd.to_numeric(df['Deaths'])
df['Death Rate'] = pd.to_numeric(df['Death Rate'])




float64


### Step 5:  Check the data's info again 

Run a `.info()` to see if the columns have missing data in them -- they should!

In [19]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1094 entries, 0 to 1093
Data columns (total 7 columns):
 #   Column                   Non-Null Count  Dtype  
---  ------                   --------------  -----  
 0   Year                     1094 non-null   int64  
 1   Leading Cause            1094 non-null   object 
 2   Sex                      1094 non-null   object 
 3   Race Ethnicity           1094 non-null   object 
 4   Deaths                   956 non-null    float64
 5   Death Rate               708 non-null    float64
 6   Age Adjusted Death Rate  1094 non-null   object 
dtypes: float64(2), int64(1), object(4)
memory usage: 60.0+ KB


## Step 6:  Create a missing inidcator (run the code below without changes)

-  This column will have a 1 if the row is missing a value in a respective column and 0 if it is not 

-  Add a comment about what the np.where() function is (feel free to use documentation)

In [20]:
#Run this cell without changes 

df['Deaths_missing'] = np.where(df['Deaths'].isna(), 1, 0)
df['Death_Rate_missing'] = np.where(df['Death Rate'].isna(), 1, 0)

### Step 7:  Calculate the median for each column 

In [24]:
deathsMedian = df['Deaths'].median()
deathRateMedian =df['Death Rate'].median()
print('Death Median:',deathsMedian,'\nDeath Rate Median:',deathRateMedian)

Death Median: 148.5 
Death Rate Median: 18.35


### Step 8:  Use the median to fill in each column's missing values (aka impute)

Hint:  Use `fillna()` with the median values you created above.

In [26]:
df['Deaths'] = df['Deaths'].fillna(deathsMedian)

df['Death Rate'] = df['Death Rate'].fillna(deathRateMedian)

In [27]:
#Check the info to see if imputation worked 

df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1094 entries, 0 to 1093
Data columns (total 9 columns):
 #   Column                   Non-Null Count  Dtype  
---  ------                   --------------  -----  
 0   Year                     1094 non-null   int64  
 1   Leading Cause            1094 non-null   object 
 2   Sex                      1094 non-null   object 
 3   Race Ethnicity           1094 non-null   object 
 4   Deaths                   1094 non-null   float64
 5   Death Rate               1094 non-null   float64
 6   Age Adjusted Death Rate  1094 non-null   object 
 7   Deaths_missing           1094 non-null   int64  
 8   Death_Rate_missing       1094 non-null   int64  
dtypes: float64(2), int64(3), object(4)
memory usage: 77.0+ KB


## Above and Beyond (AAB)  -- OPTIONAL

### Question 1:  What year had the most deaths?

In [28]:
df['Year'].value_counts()

Year
2007    141
2011    141
2010    138
2014    136
2008    136
2009    135
2012    134
2013    133
Name: count, dtype: int64

2007 and 2011 had the most deaths

### Question 2:  Change the 'Death Rate' column to a float.  

Why would you want to do this?  Add a comment answering this question.

In [29]:
df['Death Rate'] = pd.to_numeric(df['Death Rate'])
df['Death Rate'].dtypes


dtype('float64')